src/main/1.3/scala/com/holdenkarau/spark/testing/DataFrameSuiteBase.scala

```diff
@@ -109,60 +109,54 @@ trait DataFrameSuiteBaseLike extends FunSuiteLike with SparkContextProvider with
-      !(r1.isEmpty || r2.isEmpty) &&
-      !(r1.head.equals(r2.head) || DataFrameSuiteBase.approxEquals(r1.head, r2.head, 0.0))
-    }.take(1)
-    assert(unequal === List())
-    expected.rdd.unpersist()
-    result.rdd.unpersist()
-  }
 
+    try {
+      expected.rdd.cache
+      result.rdd.cache
+      assert(expected.rdd.count == result.rdd.count)
 
-  /**
-   * Zip RDD's with precise indexes. This is used so we can join two DataFrame's
-   * Rows together regardless of if the source is different but still compare based on
-   * the order.
-   */
-  private def zipWithIndex[T](input: RDD[T]): RDD[(Int, T)] = {
-    val counts = input.mapPartitions{itr => Iterator(itr.size)}.collect()
-    val countSums = counts.scanLeft(0)(_ + _).zipWithIndex.map{case (x, y) => (y, x)}.toMap
-    input.mapPartitionsWithIndex{case (idx, itr) => itr.zipWithIndex.map{case (y, i) =>
-      (i + countSums(idx), y)}
+      val expectedIndexValue = zipWithIndex(expected.rdd)
+      val resultIndexValue = zipWithIndex(result.rdd)
+
+      val unequalRDD = expectedIndexValue.join(resultIndexValue).filter{case (idx, (r1, r2)) =>
+        !(r1.equals(r2) || DataFrameSuiteBase.approxEquals(r1, r2, 0.0))}
+
+      assert(unequalRDD.take(maxUnequalRowsToShow).isEmpty)
+    } finally {
+      expected.rdd.unpersist()
+      result.rdd.unpersist()
     }
   }
 
   /**
-   * Compares if two [[DataFrame]]s are equal, checks that the schemas are the same.
-   * When comparing inexact fields uses tol.
-   */
+    * Compares if two [[DataFrame]]s are equal, checks that the schemas are the same.
+    * When comparing inexact fields uses tol.
+    *
+    * @param tol max acceptable tolerance, should be less than 1.
+    */
   def approxEqualDataFrames(expected: DataFrame, result: DataFrame, tol: Double) {
     equalSchema(expected.schema, result.schema)
-    expected.rdd.cache()
-    result.rdd.cache()
-    val expectedRDD = zipWithIndex(expected.rdd)
-    val resultRDD = zipWithIndex(result.rdd)
-    val cogrouped = expectedRDD.cogroup(resultRDD)
-    val unequal = cogrouped.filter{case (idx, (r1, r2)) =>
-      (r1.isEmpty || r2.isEmpty) || (
-        !DataFrameSuiteBase.approxEquals(r1.head, r2.head, tol))
-    }.take(1)
-    expected.rdd.unpersist()
-    result.rdd.unpersist()
-    assert(unequal === List())
+
+    try {
+      expected.rdd.cache
+      result.rdd.cache
+      assert(expected.rdd.count == result.rdd.count)
+
+      val expectedIndexValue = zipWithIndex(expected.rdd)
+      val resultIndexValue = zipWithIndex(result.rdd)
+
+      val unequalRDD = expectedIndexValue.join(resultIndexValue).filter{case (idx, (r1, r2)) =>
+        !DataFrameSuiteBase.approxEquals(r1, r2, tol)}
+
+      assert(unequalRDD.take(maxUnequalRowsToShow).isEmpty)
+    } finally {
+      expected.rdd.unpersist()
+      result.rdd.unpersist()
+    }
   }
 
+  private def zipWithIndex(rdd: RDD[Row]) = rdd.zipWithIndex().map{case (row, idx) => (idx, row)}
```

Comment:
Maybe keep the comment from the original for the use of zipWithIndex (even though its not a public method the docs are useful while working on it).